File: crl-cpp/cpp/commons/time.cpp
/**
* This file is part of project CyclingRankingsLibrary
* (https://github.com/schmouk/CyclingRankingsLib)
*
* Copyright (C) 2026 Philippe Schmouker
*
* License is GNU GENERAL PUBLIC LICENSE Version 3
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program (see:
* https://github.com/schmouk/CyclingRankingsLib/blob/main/LICENSE).
* If not, see <https://www.gnu.org/licenses/>.
*/

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <ctime>
#include <format>
#include <limits>
#include <regex>
#include <string>

#include "./time.h"


namespace crl
{
    //=====   Time Fraction of Seconds   ======================
    //---------------------------------------------------------
    SecondFraction::operator double() const noexcept
    {
        try {
            return static_cast<double>(value) / static_cast<double>(precision);
        }
        catch (...) {
            return 0.0;
        }
    }

    //---------------------------------------------------------
    SecondFraction::operator std::string() const noexcept
    {
        switch (precision) {
        case 0: case 1:
            return "";
        case 10:
            return std::format(".{}", value);
        case 100:
            return std::format(".{:02d}", value);
        case 1000:
            return std::format(".{:03d}", value);

        default:
            return std::format(" {}/{}", value, precision);
        }
    }

    //---------------------------------------------------------
    SecondFraction& SecondFraction::operator+= (const SecondFraction& other) noexcept
    {
        if (precision == other.precision) {
            value += other.value;
        }
        else {
            const std::uint32_t _common_denominator{ _lcm(precision, other.precision) };
            const std::uint32_t _numerator{
                value * (_common_denominator / precision) +
                other.value * (_common_denominator / other.precision)
            };
            const std::uint32_t gcd{ _gcd(_numerator, _common_denominator) };

            value = _numerator / gcd;
            precision = _common_denominator / gcd;
        }

        return *this;
    }

    //---------------------------------------------------------
    const bool SecondFraction::operator<(const SecondFraction& other) const noexcept
    {
        if (precision == other.precision) {
            return value < other.value;
        }
        else {
            return static_cast<double>(*this) < static_cast<double>(other) - 0.001;
        }
    }

    //---------------------------------------------------------
    const bool SecondFraction::operator==(const SecondFraction& other) const noexcept
    {
        if (precision == other.precision) {
            return value == other.value;
        }
        else {
            return std::abs(static_cast<double>(*this) - static_cast<double>(other)) < 0.001;
        }
    }

    //---------------------------------------------------------
    std::uint32_t SecondFraction::_gcd(std::uint32_t a, std::uint32_t b) noexcept
    {
        while (b != 0) {
            std::uint32_t tmp{ b };
            b = a % b;
            a = tmp;
        }
        return a;
    }

    //---------------------------------------------------------
    std::uint32_t SecondFraction::_lcm(std::uint32_t a, std::uint32_t b) noexcept
    {
        return (a / _gcd(a, b)) * b;
    }


    //=====   Time Scores   ===================================
    //---------------------------------------------------------
    Time::Time(
        const std::uint16_t h,
        const std::uint8_t  m,
        const std::uint8_t  s,
        const std::uint16_t frac_val,
        const std::uint16_t frac_prec
    ) noexcept
    {
        _evaluate_data(h, m, s, SecondFraction{ frac_val, frac_prec });
    }

    //---------------------------------------------------------
    Time::Time(
        const std::uint16_t h,
        const std::uint8_t  m,
        const std::uint8_t  s,
        const SecondFraction frac) noexcept
    {
        _evaluate_data(h, m, s, frac);
    }

    //---------------------------------------------------------
    Time::Time(
        const std::uint16_t h,
        const std::uint8_t  m,
        const std::uint8_t  s
    ) noexcept
    {
        _evaluate_data(h, m, s);
    }

    //---------------------------------------------------------
    Time::Time(const double time, const int precision) noexcept
    {
        double _int_part, _frac_part;
        _frac_part = std::modf(time, &_int_part);

        _seconds = static_cast<unsigned int>(_int_part);
        if (precision == 0) {
            _fraction = SecondFraction{ 0, 1 };
        }
        else {
            _fraction = SecondFraction{
                static_cast<std::uint16_t>(precision * _frac_part),
                static_cast<std::uint16_t>(precision)
            };
        }  
    }

    //---------------------------------------------------------
    Time::Time(const std::string& time) noexcept
    {
        _evaluate_time(time.c_str());
    }

    //---------------------------------------------------------
    Time::Time(const char* time) noexcept
    {
        if (time)
            _evaluate_time(time);
    }

    //---------------------------------------------------------
    Time& Time::operator= (std::string& time) noexcept
    {
        return *this = Time(time);
    }

    //---------------------------------------------------------
    Time& Time::operator= (const char* time) noexcept
    {
        return *this = Time(time);
    }

    //---------------------------------------------------------
    Time& Time::operator+= (const Time& other) noexcept
    {
        if (!is_ok()) {
            _error_msg = std::format("attempt to add an erroneous time ({}) -> no addition", _error_msg);
        }
        else if (other.is_ok()) {
            // Notice: this is a special operation:
            // fractions of seconds are just added together, and are not added to seconds.
            _seconds += other._seconds;
            _error_msg.clear();
            _fraction += other._fraction;
        }
        else {
            _error_msg = std::format("attempt to add an erroneous time ({}) -> no addition", other._error_msg);
        }

        return *this;
    }

    //---------------------------------------------------------
    Time& Time::operator-= (const Time& other) noexcept
    {
        if (!is_ok()) {
            _error_msg = std::format("attempt to add an erroneous time ({}) -> no substraction", _error_msg);
        }
        else if (other.is_ok()) {
            // Notice: substraction of times involves only seconds and not fractions of
            if (other._seconds <= _seconds) {
                _error_msg.clear();
                _seconds -= other._seconds;
            }
            else {
                _error_msg = std::format("can't substract {} seconds from {} seconds, result set to 0", other._seconds, _seconds);
                _seconds = 0;
            }
        }
        else {
            _error_msg = std::format("attempt to substract an erroneous time ({}) -> no substraction", other._error_msg);
        }

        return *this;
    }

    //---------------------------------------------------------
    Time::operator long() noexcept
    {
        if (_seconds <= std::numeric_limits<long>::max()) {
            return static_cast<long>(_seconds);
        }
        else {
            _error_msg = std::format(
                "{} is a too big time value to be converted to a long integer, forced to {}",
                _seconds,
                std::numeric_limits<long>::max()
            );
            return std::numeric_limits<long>::max();
        }
    }

    //---------------------------------------------------------
    Time::operator unsigned long() const noexcept
    {
        return static_cast<unsigned long>(_seconds);
    }

    //---------------------------------------------------------
    Time::operator double() const noexcept
    {
        return static_cast<double>(_seconds) + static_cast<double>(_fraction);
    }

    //---------------------------------------------------------
    Time::operator std::string() const noexcept
    {
        if (!is_ok())
            return _error_msg;

        const std::uint16_t h{ static_cast<std::uint16_t>(_seconds / 3600) };
        const std::uint8_t  m{ static_cast<std::uint8_t>((_seconds % 3600) / 60) };
        const std::uint8_t  s{ static_cast<std::uint8_t>(_seconds % 60 )};
        const std::string   frac{ std::string(_fraction) };

        if (h > 0)
            return std::format("{}:{:02d}:{:02d}{}", h, m, s, frac);
        
        if (m > 0)
            return std::format("{:d}:{:02d}{}", m, s, frac);

        return std::format("{:d}{}", s, frac);
    }

    //---------------------------------------------------------
    Time Time::operator+ (const Time& other) noexcept
    {
        Time t{ *this };
        return t += other;
    }

    //---------------------------------------------------------
    Time Time::operator- (const Time& other) noexcept
    {
        if (!is_ok())
            return *this;

        if (!other.is_ok())
            return other;

        // evaluates gaps, always gets a positive value
        const double t1{ static_cast<double>(*this) };
        const double t2{ static_cast<double>(other) };

        const int final_precision{ std::max<int>(_fraction.precision, other._fraction.precision) };

        if (t1 <= t2)
            return Time(t2 - t1, final_precision);
        else
            return Time(t1 - t2, final_precision);
    }

    //---------------------------------------------------------
    const bool Time::operator==(const Time& other) const noexcept
    {
        return _seconds == other._seconds && _fraction == other._fraction;
    }

    //---------------------------------------------------------
    const bool Time::operator!=(const Time& other) const noexcept
    {
        return !(*this == other);
    }

    //---------------------------------------------------------
    const bool Time::operator<(const Time& other) const noexcept
    {
        return (_seconds < other._seconds) ||
            (_seconds == other._seconds && _fraction < other._fraction);
    }

    //---------------------------------------------------------
    const bool Time::operator<=(const Time& other) const noexcept
    {
        return *this < other || *this < other;
    }

    //---------------------------------------------------------
    const bool Time::operator>(const Time& other) const noexcept
    {
        return !(*this <= other);
    }

    //---------------------------------------------------------
    const bool Time::operator>=(const Time& other) const noexcept
    {
        return !(*this < other);
    }

    //---------------------------------------------------------
    const bool Time::is_ok() const noexcept
    {
        return _error_msg.empty();
    }

    //---------------------------------------------------------
    const std::string& Time::get_error_message() const noexcept
    {
        return _error_msg;
    }

    //-----------------------------------------------------
    void Time::_evaluate_data(const std::uint16_t h, const std::uint8_t m, const std::uint8_t s, const SecondFraction frac) noexcept
    {
        if (m >= 60) {
            _error_msg = std::format("bad value for minutes: {}", m);
        }
        else if (s >= 60) {
            _error_msg = std::format("bad value for seconds: {}", s);
        }
        else {
            _seconds = 3600 * h + 60 * m + s;
            _fraction = frac;
            _error_msg.clear();
        }
    }
    
    //-----------------------------------------------------
    void Time::_evaluate_data(const std::uint16_t h, const std::uint8_t m, const std::uint8_t s) noexcept
    {
        _evaluate_data(h, m, s, SecondFraction{});
    }

    //-----------------------------------------------------
    void Time::_evaluate_frac(const std::string& frac_str) noexcept
    {
        if (frac_str.empty()) {
            _fraction = { 0, 1 };
            return;
        }

        const std::uint16_t frac_value{ static_cast<std::uint16_t>(std::stoi(frac_str)) };
        switch (frac_str.size()) {
        case 1:
            _fraction = { frac_value, 10 };
            break;
        case 2:
            _fraction = { frac_value, 100 };
            break;
        case 3:
            _fraction = { frac_value, 1000 };
            break;
        default:
            _fraction = { static_cast<std::uint16_t>(std::stoi(frac_str.substr(0, 3))), 1000 };
            break;
        }
    }

    //-----------------------------------------------------
    void Time::_evaluate_frac(const std::string& frac_value, const std::string& frac_precision) noexcept
    {
        _fraction.value = std::stoi(frac_value);
        _fraction.precision = std::stoi(frac_precision);
        if (_fraction.precision == 0)
            _error_msg = "bad value for precision on fractions of seconds: 0";
    }

    //-----------------------------------------------------
    void Time::_evaluate_time(const std::string& time_str) noexcept
    {
        if (_evaluate_hms_frac(time_str)  ||
            _evaluate_hms_ratio(time_str) ||
            _evaluate_ms_frac(time_str)   ||
            _evaluate_s_frac(time_str)    ||
            _evaluate_ms_ratio(time_str)  ||
            _evaluate_s_ratio(time_str)     )
        {
            // some match found, things are fine
            return;
        }

        // otherwise: no match found, form of time is erroneous
        _error_msg = std::format("erroneous format for time string : '{}'", time_str);
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_hms_ratio(const std::string& time_str) noexcept
    {
        // HHH:MM:SS frac_val/frac_precision
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d*):(\\d\\d):(\\d\\d) (\\d+)/(\\d+)")))
        {
            // Match h, m, s and fraction!
            const std::int32_t h{ std::stol(time_matches[1]) };
            const std::int32_t m{ std::stol(time_matches[2]) };
            const std::int32_t s{ std::stol(time_matches[3]) };
            _seconds = 3600 * h + 60 * m + s;

            _evaluate_frac(time_matches[4], time_matches[5]);
            return true;
        }

        return false;
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_hms_frac(const std::string& time_str) noexcept
    {
        // HHH:MM:SS.frac
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d*):(\\d\\d):(\\d\\d)(\\.(\\d+))?")))
        {
            // Match h, m, s and maybe fraction!
            const std::int32_t h{ std::stol(time_matches[1]) };
            const std::int32_t m{ std::stol(time_matches[2]) };
            const std::int32_t s{ std::stol(time_matches[3]) };
            _seconds = 3600 * h + 60 * m + s;

            _evaluate_frac(time_matches[5]);
            return true;
        }

        return false;
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_ms_ratio(const std::string& time_str) noexcept
    {
        // MM:SS frac_val/frac_precision
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d?):(\\d\\d) (\\d+)/(\\d+)")))
        {
            // Match m, s and fraction!
            const std::int32_t m{ std::stol(time_matches[1]) };
            const std::int32_t s{ std::stol(time_matches[2]) };
            _seconds = 60 * m + s;

            _evaluate_frac(time_matches[3], time_matches[4]);
            return true;
        }

        return false;
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_ms_frac(const std::string& time_str) noexcept
    {
        // MM:SS.frac
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d*):(\\d\\d)(\\.(\\d+))?")))
        {
            // Match m, s and maybe fraction!
            const std::int32_t m{ std::stol(time_matches[1]) };
            const std::int32_t s{ std::stol(time_matches[2]) };
            _seconds = 60 * m + s;

            _evaluate_frac(time_matches[4]);
            return true;
        }

        return false;
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_s_ratio(const std::string& time_str) noexcept
    {
        // SS frac_val/frac_precision
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d?) (\\d+)/(\\d+)")))
        {
            // Match s and fraction!
            _seconds = std::stol(time_matches[1]);
            _evaluate_frac(time_matches[2], time_matches[3]);
            return true;
        }

        return false;
    }

    //---------------------------------------------------------
    const bool Time::_evaluate_s_frac(const std::string& time_str) noexcept
    {
        // SS.frac
        std::smatch time_matches;
        _error_msg.clear();

        if (std::regex_search(time_str, time_matches,
            std::regex("^(\\d\\d?)(\\.(\\d+))?")))
        {
            // Match s and maybe fraction!
            _seconds = std::stol(time_matches[1]);
            _evaluate_frac(time_matches[3]);
            return true;
        }

        return false;
    }


    //=====   HMTime   ========================================
    //---------------------------------------------------------
    HMTime::HMTime(const std::uint16_t h, const std::uint8_t m) noexcept
        : Time{ h, m, 0 }
    {}

    //---------------------------------------------------------
    HMTime::HMTime(const std::string& time) noexcept
        : Time{}
    {
        _evaluate_time(time);
    }

    //---------------------------------------------------------
    HMTime::HMTime(const char* time) noexcept
        : Time{}
    {
        _evaluate_time(std::string(time));
    }

    //---------------------------------------------------------
    void HMTime::_evaluate_time(const std::string& str) noexcept
    {
        // HHH:MM
        std::smatch time_matches;

        if (std::regex_search(str, time_matches, std::regex("^(\\d\\d*):(\\d\\d)"))) {
            // Match h, m found!
            const std::int32_t h{ std::stol(time_matches[1]) };
            const std::int32_t m{ std::stol(time_matches[2]) };
            _seconds = 3600 * h + 60 * m;

            _error_msg.clear();
        }
        else {
            _error_msg = std::format("erroneous format for 'hours:minutes' time string: '{}'", str);
        }
    };


    //=====   MSTime   ========================================
    //---------------------------------------------------------
    MSTime::MSTime(const std::uint8_t m, const std::uint8_t s, const std::uint16_t frac_val, const std::uint16_t frac_prec) noexcept
        : Time{ 0, m, s, frac_val, frac_prec }
    {}

    //---------------------------------------------------------
    MSTime::MSTime(const std::uint8_t m, const std::uint8_t s, const SecondFraction frac) noexcept
        : Time{ 0, m, s, frac }
    {}

    //---------------------------------------------------------
    MSTime::MSTime(const std::uint8_t m, const std::uint8_t s) noexcept
        : Time{ 0, m, s }
    {}

    //---------------------------------------------------------
    MSTime::MSTime(const double time, const int precision) noexcept
        : Time{ time, precision }
    {}

    //---------------------------------------------------------
    MSTime::MSTime(const std::string& time) noexcept
        : Time{ time }
    {}

    //---------------------------------------------------------
    MSTime::MSTime(const char* time) noexcept
        : Time{ time }
    {}

    //---------------------------------------------------------
    void MSTime::_evaluate_time(const std::string& str) noexcept
    {
        if (_evaluate_ms_frac(str) || _evaluate_ms_ratio(str)) {
            _error_msg.clear();
        }
        else {
            _error_msg = std::format("erroneous format for 'minutes:seconds + fraction' time string: '{}'", str);
        }
    }


    //=====   STime   =========================================
    //---------------------------------------------------------
    STime::STime(const std::uint8_t s, const std::uint16_t frac_val, const std::uint16_t frac_prec) noexcept
        : Time{ 0, 0, s, frac_val, frac_prec }
    {}

    //---------------------------------------------------------
    STime::STime(const std::uint8_t s, const SecondFraction frac) noexcept
        : Time{ 0, 0, s, frac }
    {}

    //---------------------------------------------------------
    STime::STime(const unsigned int s) noexcept
        : Time{ static_cast<double>(s) }
    {}

    //---------------------------------------------------------
    STime::STime(const double time, const int precision) noexcept
        : Time{ time, precision }
    {}

    //---------------------------------------------------------
    STime::STime(const std::string& time) noexcept
        : Time{ time }
    {}

    //---------------------------------------------------------
    STime::STime(const char* time) noexcept
        : Time{ time }
    {}

    //---------------------------------------------------------
    void STime::_evaluate_time(const std::string& str) noexcept
    {
        if (_evaluate_s_frac(str) || _evaluate_s_ratio(str)) {
            _error_msg.clear();
        }
        else {
            _error_msg = std::format("erroneous format for 'seconds + fraction' time string: '{}'", str);

        }
    }

}

File: crl-python/commons/time.py
"""
This file is part of project CyclingRankingsLibrary
(https://github.com/schmouk/CyclingRankingsLib)

copyright (c) 2026 Philippe Schmouker

License is GNU GENERAL PUBLIC LICENSE Version 3
(full text available at:
 https://github.com/schmouk/CyclingRankingsLib?tab=GPL-3.0-1-ov-file)
"""

import re
import sys
from typing import Optional, Union


#======   Time Fraction of Seconds   ==========================
class SecondFraction:
    """Represents a fraction of seconds for time calculations."""

    #----------------------------------------------------------
    def __init__(self, value: int = 0, precision: int = 1) -> None:
        """Initialize a SecondFraction.

        Args:
            value: Numerator of the fraction (may be greater than precision)
            precision: Denominator of the fraction (mostly 5, 10, 100 or 1000)
        """
        self.value: int = value
        self.precision: int = precision

    #----------------------------------------------------------
    def __float__(self) -> float:
        """Convert fraction to float."""
        try:
            return float(self.value) / float(self.precision)
        except:
            return 0.0

    #----------------------------------------------------------
    def __str__(self) -> str:
        """Convert fraction to string representation."""
        match self.precision:
            case 0 | 1:
                return ""
            case 10:
                return f".{self.value}"
            case 100:
                return f".{self.value:02d}"
            case 1000:
                return f".{self.value:03d}"
            case _:
                return f" {self.value}/{self.precision}"

    #----------------------------------------------------------
    def __iadd__(self, other: 'SecondFraction') -> 'SecondFraction':
        """Add another fraction to this fraction (+=)."""
        if self.precision == other.precision:
            self.value += other.value
        else:
            common_denominator: int = self._lcm(self.precision, other.precision)
            numerator: int = (
                self.value * (common_denominator // self.precision) +
                other.value * (common_denominator // other.precision)
            )
            gcd: int = self._gcd(numerator, common_denominator)
            self.value = numerator // gcd
            self.precision = common_denominator // gcd

        return self

    #----------------------------------------------------------
    @staticmethod
    def _gcd(a: int, b: int) -> int:
        """Calculate greatest common divisor."""
        while b != 0:
            a, b = b, a % b
        return a

    #----------------------------------------------------------
    @staticmethod
    def _lcm(a: int, b: int) -> int:
        """Calculate least common multiple."""
        return (a // SecondFraction._gcd(a, b)) * b


#=====   Time Scores - Base class   ===========================
class Time:
    """Represents a time value with seconds and fractions.
    This is the base class for all other TimeXYZ classes.
    """

    #----------------------------------------------------------
    def __init__(
        self,
        h: Optional[int] = None,
        m: Optional[int] = None,
        s: Optional[int] = None,
        frac_val: Optional[int | SecondFraction] = None,
        frac_prec: Optional[int] = None
    ) -> None:
        """Initialize a Time object."""
        # Determines which constructor pattern was used
        if h is None:
            # Empty constructor
            self._seconds: int = 0
            self._fraction: SecondFraction = SecondFraction(0, 1)
            self._error_msg: str = ""

        elif m is None or s is None:
            # Erroneous constructor
            self._seconds: int = 0
            self._fraction: SecondFraction = SecondFraction(0, 1)
            self._error_msg: str = f"Erroneous format for creation of a Time: ({h}:{m}:{s})"

        else:
            self._evaluate_data(h, m, s, frac_val, frac_prec)

    #----------------------------------------------------------
    def __int__(self) -> int:
        """Convert to integer (long)."""
        if self._seconds <= sys.maxsize:
            return int(self._seconds)
        else:
            self._error_msg = (
                f"{self._seconds} is a too big time value to be converted to a "
                f"long integer, forced to {sys.maxsize}"
            )
            return sys.maxsize

    #----------------------------------------------------------
    def __float__(self) -> float:
        """Convert to float."""
        return float(self._seconds) + float(self._fraction)

    #----------------------------------------------------------
    def __str__(self) -> str:
        """Convert to string representation."""
        if not self.is_ok():
            return self._error_msg

        h = self._seconds // 3600
        m = (self._seconds % 3600) // 60
        s = self._seconds % 60
        frac = str(self._fraction)

        if h > 0:
            return f"{h}:{m:02d}:{s:02d}{frac}"

        if m > 0:
            return f"{m:2d}:{s:02d}{frac}"

        return f"{s:2d}{frac}"

    #----------------------------------------------------------
    def __add__(self, other: 'Time') -> 'Time':
        """Add two times."""
        result = Time()
        result._seconds = self._seconds + other._seconds
        result._fraction = SecondFraction(self._fraction.value, self._fraction.precision)
        result._fraction += other._fraction
        result._error_msg = self._error_msg if self._error_msg else other._error_msg
        return result

    #----------------------------------------------------------
    def __sub__(self, other: 'Time') -> 'Time':
        """Subtract two times (evaluates gaps, always positive)."""
        if not self.is_ok():
            return Time()  # Return empty Time

        if not other.is_ok():
            return Time()  # Return empty Time

        t1 = float(self)
        t2 = float(other)

        final_precision = max(self._fraction.precision, other._fraction.precision)

        if t1 <= t2:
            return Time(t2 - t1, final_precision)
        else:
            return Time(t1 - t2, final_precision)

    #----------------------------------------------------------
    def __iadd__(self, other: 'Time') -> 'Time':
        """Add another time to this time (+=)."""
        if not self.is_ok():
            self._error_msg = f"attempt to add an erroneous time ({self._error_msg}) -> no addition"
        elif other.is_ok():
            # Special operation: fractions are added, but not to seconds
            self._seconds += other._seconds
            self._error_msg = ""
            self._fraction += other._fraction
        else:
            self._error_msg = f"attempt to add an erroneous time ({other._error_msg}) -> no addition"

        return self

    #----------------------------------------------------------
    def __isub__(self, other: 'Time') -> 'Time':
        """Subtract another time from this time (-=)."""
        if not self.is_ok():
            self._error_msg = f"attempt to add an erroneous time ({self._error_msg}) -> no substraction"
        elif other.is_ok():
            # Substraction involves only seconds, not fractions
            if other._seconds <= self._seconds:
                self._error_msg = ""
                self._seconds -= other._seconds
            else:
                self._error_msg = (
                    f"can't substract {other._seconds} seconds from {self._seconds} "
                    f"seconds, result set to 0"
                )
                self._seconds = 0
        else:
            self._error_msg = f"attempt to substract an erroneous time ({other._error_msg}) -> no substraction"

        return self

    #----------------------------------------------------------
    def is_ok(self) -> bool:
        """Check if this Time object is valid (no error)."""
        return len(self._error_msg) == 0

    #----------------------------------------------------------
    def get_error_message(self) -> str:
        """Get the error message if any."""
        return self._error_msg

    #----------------------------------------------------------
    def _evaluate_data(
        self,
        h: int,
        m: int,
        s: int,
        frac_val: Optional[int | SecondFraction] = None,
        frac_prec: Optional[int] = None
    ) -> None:
        """Evaluate and set time data from h, m, s and optional fraction."""
        if m >= 60:
            self._error_msg = f"bad value for minutes: {m}"
            return
        
        if s >= 60:
            self._error_msg = f"bad value for seconds: {s}"
            return

        self._seconds = 3600 * h + 60 * m + s
        if frac_val is None:
            self._fraction = SecondFraction(0, 1)
        else:
            self._fraction = SecondFraction(frac_val, int(frac_prec)) if isinstance(frac_val, int) else frac_val
        
        if self._fraction.precision == 0:
            self._error_msg = "bad value for precision on fractions of seconds: 0"
        else:
            self._error_msg = ""

    #----------------------------------------------------------
    def _evaluate_frac(self, frac_str: str, frac_precision: Optional[str] = None) -> None:
        """Evaluate fraction from string representation."""
        self._error_msg = ""

        if not frac_str:
            self._fraction = SecondFraction(0, 1)
            return

        if frac_precision is not None:
            # Pattern: "value/precision"
            self._fraction.value = int(frac_str)
            self._fraction.precision = int(frac_precision)
            if self._fraction.precision == 0:
                self._error_msg = "bad value for precision on fractions of seconds: 0"

        else:
            # Pattern: ".digits"
            frac_value = int(frac_str)
            match len(frac_str):
                case 1:
                    self._fraction = SecondFraction(frac_value, 10)
                case 2:
                    self._fraction = SecondFraction(frac_value, 100)
                case 3:
                    self._fraction = SecondFraction(frac_value, 1000)
                case _:
                    self._fraction = SecondFraction(int(frac_str[:3]), 1000)


#=====   Time Scores - HMS class   ============================
class HMSTime(Time):
    #----------------------------------------------------------
    def __init__(
        self,
        h: Optional[int] = None,
        m: Optional[int] = None,
        s: Optional[int] = None,
        frac_val: Optional[int | SecondFraction] = None,
        frac_prec: Optional[int] = None
    ) -> None:
        super().__init__(h, m, s, frac_val, frac_prec)


#=====   Time Scores - MS class   =============================
class MSTime(Time):
    #----------------------------------------------------------
    def __init__(
        self,
        m: Optional[int] = None,
        s: Optional[int] = None,
        frac_val: Optional[int | SecondFraction] = None,
        frac_prec: Optional[int] = None
    ) -> None:
        super().__init__(0, m, s, frac_val, frac_prec)


#=====   Time Scores - S class   ==============================
class STime(Time):
    #----------------------------------------------------------
    def __init__(
        self,
        s: Optional[int] = None,
        frac_val: Optional[int | SecondFraction] = None,
        frac_prec: Optional[int] = None
    ) -> None:
        super().__init__(0, 0, s, frac_val, frac_prec)


#=====   Time Scores - creation from double value   ===========
class DoubleTime(Time):
    #----------------------------------------------------------
    def __init__(self, time_val: float, precision: Optional[int] = None) -> None:
        """Initialize from a double value with maybe precision."""
        super().__init__()

        int_part = int(time_val)
        frac_part = time_val - int_part

        self._seconds = int_part
        if precision is None:
            self._fraction = SecondFraction(0, 1)
        else:
            self._fraction = SecondFraction(
                int(precision * frac_part),
                precision
            )
         
            
#=====   Time Scores - creation from str value   ==============
class StrTime(Time):
#----------------------------------------------------------
    def __init__(self, time_str: str) -> None:
        """Evaluate time from string representation.

        Supports formats:
        - HHH:MM:SS frac_val/frac_precision
        - HHH:MM:SS.frac
        - MM:SS frac_val/frac_precision
        - MM:SS.frac
        - SS frac_val/frac_precision
        - SS.frac
        """
        super().__init__()

        # HHH:MM:SS frac_val/frac_precision
        match = re.search(r'^(\d\d*):(\d\d):(\d\d) (\d+)/(\d+)', time_str)
        if match:
            h = int(match.group(1))
            m = int(match.group(2))
            s = int(match.group(3))
            self._seconds = 3600 * h + 60 * m + s
            self._evaluate_frac(match.group(4), match.group(5))
            return

        # HHH:MM:SS.frac
        match = re.search(r'^(\d\d*):(\d\d):(\d\d)(\.(\d+))?', time_str)
        if match:
            h = int(match.group(1))
            m = int(match.group(2))
            s = int(match.group(3))
            self._seconds = 3600 * h + 60 * m + s
            self._evaluate_frac(match.group(5) if match.group(5) else "")
            return

        # MM:SS frac_val/frac_precision
        match = re.search(r'^(\d\d?):(\d\d) (\d+)/(\d+)', time_str)
        if match:
            m = int(match.group(1))
            s = int(match.group(2))
            self._seconds = 60 * m + s
            self._evaluate_frac(match.group(3), match.group(4))
            return

        # MM:SS.frac
        match = re.search(r'^(\d\d*):(\d\d)(\.(\d+))?', time_str)
        if match:
            m = int(match.group(1))
            s = int(match.group(2))
            self._seconds = 60 * m + s
            self._evaluate_frac(match.group(4) if match.group(4) else "")
            return

        # SS frac_val/frac_precision
        match = re.search(r'^(\d\d?) (\d+)/(\d+)', time_str)
        if match:
            s = int(match.group(1))
            self._seconds = s
            self._evaluate_frac(match.group(2), match.group(3))
            return

        # SS.frac
        match = re.search(r'^(\d\d?)(\.(\d+))?', time_str)
        if match:
            s = int(match.group(1))
            self._seconds = s
            self._evaluate_frac(match.group(3) if match.group(3) else "")
            return
